### DISCUSSION TITLE
Adding Pset_Address to a building

### DISCUSSION THREAD

**Discussion by Booplicate** [IfcOpenShell/IfcOpenShell#3228] [Q&A]
I'm experimenting on generating ifc models, currently I'm stuck at adding an address to a building, this is what I'm trying to do:
```py
pset = run("pset.add_pset", model, product=building, name="Pset_Address")
run(
    "pset.edit_pset",
    model,
    pset=pset,
    properties={
        "Country": "Test Country",
        "Region": "Test Region",
        "Town": "Test City",
        "AddressLines": ("foo", "bar", "egg")
    }
)
```
When `AddressLines` is a tuple, I'm getting `RecursionError: maximum recursion depth exceeded in comparison`, when it's a list, I'm getting `AttributeError: 'NoneType' object has no attribute 'HasPropertyTemplates'`. If I set it to a string, it works, but doesn't appear to be correct according to building smart, correct me if I'm wrong.

So my question is, am I doing it wrong, or is there a bug? Or maybe I must define my own `pset_template`?

**Comment by aothms** [comment]
I tried various things, but I couldn't get a RecursionError to pop up. Can you maybe share full working code?

```
>>> import ifcopenshell
>>>
>>> from ifcopenshell.api import run
>>>
>>> f = ifcopenshell.file(schema='IFC4X3')
>>> b = f.createIfcBuilding()
>>>
>>> pset = run("pset.add_pset", f, product=b, name="Pset_Address")
>>> run(
...     "pset.edit_pset",
...     f,
...     pset=pset,
...     properties={
...         "Country": "Test Country",
...         "Region": "Test Region",
...         "Town": "Test City",
...         "AddressLines": ("foo", "bar", "egg")
...     }
... )
>>>
>>>
>>> pset
#2=IfcPropertySet('3fjoM2BBD85eABRK6krOyu',$,'Pset_Address',$,(#4,#5,#6,#7))
>>> pset.HasProperties
(#4=IfcPropertySingleValue('Country',$,IfcLabel('Test Country'),$), #5=IfcPropertySingleValue('Region',$,IfcLabel('Test Region'),$), #6=IfcPropertySingleValue('Town',$,IfcLabel('Test City'),$), #7=IfcPropertySingleValue('AddressLines',$,IfcLabel('('foo', 'bar', 'egg')'),$))
```

In my test the tuple get serialized into a string as a single label.

Looking at the code around here https://github.com/IfcOpenShell/IfcOpenShell/blob/v0.7.0/src/ifcopenshell-python/ifcopenshell/api/pset/edit_pset.py#L258 there will be various things that need changing, like testing for (tuple, list) instead of just list, and more importantly supporting IfcPropertyListValue.

Is this something you'd perhaps like to contribute to?

**Reply by Booplicate** [reply]
I see, the tuple error happened because I didn't specify the schema version, makes sense, it must be `IFC4X3` since `AddressLines` is a new pset.
The list doesn't work because it expects the value to be some kind of enumeration and there's no support for `IfcPropertyListValue` yet.

So `pset_template.Enumerators` is `None` for me (looking back at `AttributeError: 'NoneType' object has no attribute 'HasPropertyTemplates`) because it's not an enum, but a list, I haven't dealt with templates before, so will have to see what I get in the case of `IfcPropertyListValue`.


I tried to manually create an instance of `IfcPropertyListValue` via
```py
address_lines = run("root.create_entity", model, ifc_class="IfcPropertyListValue")
```
But I got a new error: `IndexError: list index out of range` I then tried to create an entity another way:
```py
address_lines = model.create_entity("IfcPropertyListValue")
```
and this time it worked. Not sure why I couldn't create it via `run`.
Then I manually populated the list and assigned it to pset. The final ifc file *seems* to be correct, but I'm not an ifc expert.
```py
address_lines.ListValues = [
    model.create_entity("IfcLabel", "foo"),
    model.create_entity("IfcLabel", "bar"),
    model.create_entity("IfcLabel", "egg")
]
run(
    "pset.edit_pset",
    model,
    pset=pset,
    properties={
        "Country": "Test Country",
        "Region": "Test Region",
        "Town": "Test City",
        "AddressLines": address_lines
    }
)
```

@aothms thanks for guiding me in the right direction. I definitely would like to contribute in the future, right now my understanding of ifc and this lib may not be enough, but if I figure out a way to make `IfcPropertyListValue` work directly there, I will submit a PR.

1. Need to learn how to use templates
2. Change that `if` check to handle both list and tuple
3. Somehow distinguish when the passed value is an enumeration and when it's a list
    - then it'd be something like I did: create `IfcPropertyListValue` and populate it with the values. Not sure how to choose the correct ifc class for the values, it won't always be `IfcLabel`
4. Find out why I couldn't create `IfcPropertyListValue` with `run`

**Reply by Booplicate** [reply]
Hmm, but when I'm trying to read the new psets, I'm getting an error:
```py
psets = get_psets(building, psets_only=True)
```
`AttributeError: entity instance of type 'IFC4X3.IfcPropertyListValue' has no attribute 'wrappedValue'`

UPD:
Found, it's also an issue with `add_new_properties`, it only [supports enumeration](https://github.com/IfcOpenShell/IfcOpenShell/blob/1488ae40eeacb8d7226268c639084625ec9f1f49/src/ifcopenshell-python/ifcopenshell/api/pset/edit_pset.py#L246-L256C22), so my list is actually stored as `IfcPropertySingleValue` while being `IfcPropertyListValue`. I wonder what would be the best approach to fix it? I don't think adding another `elif` branch would help.

**Reply by aothms** [reply]
Yep, the `else`-branch expects a simple type instance, like IfcLabel.

I think adding the `elif` is the way forward. There aren't that many property types.

Alternatively we restructure the statement altogether and really check whether the type given is actually a data type that is acceptable as an IfcPropertySingleValue.NominalValue (a type named IfcValue) or that it is a subtype of property definition 

```
>>> import ifcopenshell
>>> ifcopenshell.ifcopenshell_wrapper.schema_by_name('ifc4').declaration_by_name('ifcvalue')
<select IfcValue: (<select IfcDerivedMeasureValue: (<type IfcAbsorbedDoseMeasure: <real>> | <type IfcAccelerationMeasure: <real>> | <type IfcAngularVelocityMeasure: <real>> | <type IfcAreaDensityMeasure: <real>> | <type IfcCompoundPlaneAngleMeasure: <list [3:4] of <integer>>> | <type IfcCurvatureMeasure: <real>> | <type IfcDoseEquivalentMeasure: <real>> | <type IfcDynamicViscosityMeasure: <real>> | <type IfcElectricCapacitanceMeasure: <real>> | <type IfcElectricChargeMeasure: <real>> | <type IfcElectricConductanceMeasure: <real>> | <type IfcElectricResistanceMeasure: <real>> | <type IfcElectricVoltageMeasure: <real>> | <type IfcEnergyMeasure: <real>> | <type IfcForceMeasure: <real>> | <type IfcFrequencyMeasure: <real>> | <type IfcHeatFluxDensityMeasure: <real>> | <type IfcHeatingValueMeasure: <real>> | <type IfcIlluminanceMeasure: <real>> | <type IfcInductanceMeasure: <real>> | <type IfcIntegerCountRateMeasure: <integer>> | <type IfcIonConcentrationMeasure: <real>> | <type IfcIsothermalMoistureCapacityMeasure: <real>> | <type IfcKinematicViscosityMeasure: <real>> | <type IfcLinearForceMeasure: <real>> | <type IfcLinearMomentMeasure: <real>> | <type IfcLinearStiffnessMeasure: <real>> | <type IfcLinearVelocityMeasure: <real>> | <type IfcLuminousFluxMeasure: <real>> | <type IfcLuminousIntensityDistributionMeasure: <real>> | <type IfcMagneticFluxDensityMeasure: <real>> | <type IfcMagneticFluxMeasure: <real>> | <type IfcMassDensityMeasure: <real>> | <type IfcMassFlowRateMeasure: <real>> | <type IfcMassPerLengthMeasure: <real>> | <type IfcModulusOfElasticityMeasure: <real>> | <type IfcModulusOfLinearSubgradeReactionMeasure: <real>> | <type IfcModulusOfRotationalSubgradeReactionMeasure: <real>> | <type IfcModulusOfSubgradeReactionMeasure: <real>> | <type IfcMoistureDiffusivityMeasure: <real>> | <type IfcMolecularWeightMeasure: <real>> | <type IfcMomentOfInertiaMeasure: <real>> | <type IfcMonetaryMeasure: <real>> | <type IfcPHMeasure: <real>> | <type IfcPlanarForceMeasure: <real>> | <type IfcPowerMeasure: <real>> | <type IfcPressureMeasure: <real>> | <type IfcRadioActivityMeasure: <real>> | <type IfcRotationalFrequencyMeasure: <real>> | <type IfcRotationalMassMeasure: <real>> | <type IfcRotationalStiffnessMeasure: <real>> | <type IfcSectionModulusMeasure: <real>> | <type IfcSectionalAreaIntegralMeasure: <real>> | <type IfcShearModulusMeasure: <real>> | <type IfcSoundPowerLevelMeasure: <real>> | <type IfcSoundPowerMeasure: <real>> | <type IfcSoundPressureLevelMeasure: <real>> | <type IfcSoundPressureMeasure: <real>> | <type IfcSpecificHeatCapacityMeasure: <real>> | <type IfcTemperatureGradientMeasure: <real>> | <type IfcTemperatureRateOfChangeMeasure: <real>> | <type IfcThermalAdmittanceMeasure: <real>> | <type IfcThermalConductivityMeasure: <real>> | <type IfcThermalExpansionCoefficientMeasure: <real>> | <type IfcThermalResistanceMeasure: <real>> | <type IfcThermalTransmittanceMeasure: <real>> | <type IfcTorqueMeasure: <real>> | <type IfcVaporPermeabilityMeasure: <real>> | <type IfcVolumetricFlowRateMeasure: <real>> | <type IfcWarpingConstantMeasure: <real>> | <type IfcWarpingMomentMeasure: <real>>)> | <select IfcMeasureValue: (<type IfcAmountOfSubstanceMeasure: <real>> | <type IfcAreaMeasure: <real>> | <type IfcComplexNumber: <array [1:2] of <real>>> | <type IfcContextDependentMeasure: <real>> | <type IfcCountMeasure: <number>> | <type IfcDescriptiveMeasure: <string>> | <type IfcElectricCurrentMeasure: <real>> | <type IfcLengthMeasure: <real>> | <type IfcLuminousIntensityMeasure: <real>> | <type IfcMassMeasure: <real>> | <type IfcNonNegativeLengthMeasure: <type IfcLengthMeasure: <real>>> | <type IfcNormalisedRatioMeasure: <type IfcRatioMeasure: <real>>> | <type IfcNumericMeasure: <number>> | <type IfcParameterValue: <real>> | <type IfcPlaneAngleMeasure: <real>> | <type IfcPositiveLengthMeasure: <type IfcLengthMeasure: <real>>> | <type IfcPositivePlaneAngleMeasure: <type IfcPlaneAngleMeasure: <real>>> | <type IfcPositiveRatioMeasure: <type IfcRatioMeasure: <real>>> | <type IfcRatioMeasure: <real>> | <type IfcSolidAngleMeasure: <real>> | <type IfcThermodynamicTemperatureMeasure: <real>> | <type IfcTimeMeasure: <real>> | <type IfcVolumeMeasure: <real>>)> | <select IfcSimpleValue: (<type IfcBinary: <binary>> | <type IfcBoolean: <boolean>> | <type IfcDate: <string>> | <type IfcDateTime: <string>> | <type IfcDuration: <string>> | <type IfcIdentifier: <string>> | <type IfcInteger: <integer>> | <type IfcLabel: <string>> | <type IfcLogical: <logical>> | <type IfcPositiveInteger: <type IfcInteger: <integer>>> | <type IfcReal: <real>> | <type IfcText: <string>> | <type IfcTime: <string>> | <type IfcTimeStamp: <integer>>)>)>
```

```
>>> f = ifcopenshell.file()
>>> f.createIfcPropertyListValue().is_a('IfcProperty')
True
>>> f.createIfcPropertySingleValue().is_a('IfcProperty')
True
>>> f.createIfcPropertyEnumeratedValue().is_a('IfcProperty')
True
```

At least indeed in get_psets() IfcPropertyListValue seems properly handled

https://github.com/IfcOpenShell/IfcOpenShell/blob/fa4e19eba7db7d163c4c0c80df6443227e1e3095/src/ifcopenshell-python/ifcopenshell/util/element.py#L219-L220

**Reply by Booplicate** [reply]
Let me rephrase to make sure I got it right, since I'm new to ifc :)

If I understand your idea correctly, we end up with 3 paths:
1. If `value` is an `IfcProperty` or its subclass, then we just append it to the properties
2. If `value` is an `IfcValue`, then we create `IfcPropertySingleValue`, since `NominalValue` must be `IfcValue`
3. Something is wrong and we raise an error?

(could also use `.is_entity()`, but it's less flexible)

If that is your idea, then I like it more, and it sounds easy enough to do even for me heh.

**Comment by aothms** [comment]
> Not sure why I couldn't create it via run.

Yes this is somewhat unfortunate naming, but api.create_entity() is not intended for arbitrary classes, see the note here: https://github.com/IfcOpenShell/IfcOpenShell/blob/v0.7.0/src/ifcopenshell-python/ifcopenshell/api/root/create_entity.py#L27-L29

> Somehow distinguish when the passed value is an enumeration and when it's a list

This is exactly what the template provides. It basically represents the page you see here https://ifc43-docs.standards.buildingsmart.org/IFC/RELEASE/IFC4x3/HTML/lexical/Pset_Address.htm with 

Name | Property Type | Data Type | Description
-- | -- | -- | --
AddressLines | IfcPropertyListValue | IfcLabel | The postal address.

i.e IfcPropertyListValue of IfcLabel in computer interpretable form. See below:

```
>>> import ifcopenshell.util.pset
>>> pq = ifcopenshell.util.pset.get_template('IFC4X3')
>>> [y for y in [x for x in pq.templates[0].by_type('IfcPropertySetTemplate') if x.Name == 'Pset_Address'][0].HasPropertyTemplates if y.Name == 'AddressLines'][0]
#47=IfcSimplePropertyTemplate('0luNvvEdv0SPOspsLw3EAo',$,'AddressLines','The postal address.NOTE  A postal address may occupy several lines (or elements) when recorded. It is expected that normal usage will incorporate relevant elements of the following address concepts: A location within a building (e.g. 3rd Floor) Building name (e.g. Interoperability House) Street number (e.g. 6400) Street name (e.g. Alliance Boulevard). Typical content of address lines may vary in different countries.',.P_LISTVALUE.,$,$,$,$,$,$,.READWRITE.)
>>> [y for y in [x for x in pq.templates[0].by_type('IfcPropertySetTemplate') if x.Name == 'Pset_Address'][0].HasPropertyTemplates if y.Name == 'AddressLines'][0].TemplateType
'P_LISTVALUE'
```
